### ISSUE TITLE
I failed to make use of the gem without inspecting the source code with a debugger

### ISSUE THREAD

**Issue by x-yuri** [mhenrixon/sidekiq-unique-jobs#533]
> The goal of this gem is to ensure your Sidekiq jobs are unique. We do this by creating unique keys in Redis based on how you configure uniqueness.

Okay, you probably *can* describe the purpose in terms of uniqueness. But in order to use the gem the user must understand locks and strategies. So, I'd say:

> This gem lets you decide how to go about duplicate jobs. Duplicate jobs are jobs that have the same class, queue, and arguments. It does so by obtaining locks on digests of those 3 things (class, queue, arguments). What happens when a lock can't be obtained is defined by a chosen strategy.

Then goes General Information, Global Configuration, Worker Configuration, Locks, Conflict Strateg<strike>y</strike>ies, and only then Usage. Do you expect user to read and understand all of that before starting using the gem. I believe Usage should go before General Information (Installation, Support Me, Usage), and be like this:

> Require the gem:
> 
>     require 'sidekiq-unique-jobs'
> 
> Specify a lock type, and optionally a strategy:
> 
> ```rb
> class HardWorker
>   include Sidekiq::Worker
>   sidekiq_options lock: :while_executed, strategy: :reject
>   ...
> ```
> 
> The builtin lock types:
> 
> * `until_executing`, obtained before adding to a queue, released before executing the job
> * `until_executed`, obtained before adding to a queue, released after the job has finished executing
> * `until_expired`, obtained before adding to a queue, released after the specified timeout has been reached
> * `while_executing`, obtained before executing the job, released after the job has finished executing
> * `until_and_while_executing`, like `until_executed`, but before executing the job the lock is reobtained (`unlock` + `lock`)
> 
> The builtin strategies are: `raise`, `reject`, `replace`, `reschedule`, `log`, `null` (the default).

P.S. What's with your [table of contents][1]? Do you use some TOC generator?

[1]: https://github.com/mhenrixon/sidekiq-unique-jobs/commit/fe6dcf52a5f112f1a3a04148c44c2f559d502b44#diff-04c6e90faac2675aa89e2176d2eec7d8R36

**Comment by mhenrixon** [comment]
First of all, let me just say that I am aware of the problem. I suck at writing comprehensible and user-friendly documentation. 

That said, it is intended that v7 should be released with an improved guide where the details are kind of linked to a wiki page for the gory details of how things work and the main README should more or less be a guide on how to get everything setup for a few select use cases. 

This gem is complex and due to how difficult it has been to get it working perfectly I have postponed this until I am ready with version 7 (a version that just works).

Until then, I indeed want people to truly come to terms with how the gem works internally so that they can help me get it ready. I basically want it to be on par with sidekiq pro uniqueness for v7.

Your suggestion is great though, I sincerely appreciate any guidance on how to improve the documentation.

To answer your last comment I am using a plugin for Sublime Text that is called MarkdownTOC which updates the TOC every time I save the README.md.

**Comment by x-yuri** [comment]
My main concern with the TOC was that it was broken, but as I see you've already fixed it.

Regarding having parts of the documentation in the wiki, ideally it should all be versioned. Like you know, when you need to fix a bug in some old site, and suddenly you realize that you've got no documentation for some gem.

But I totally understand you wanting to lower the burden of maintaining the documentation. So being able to find some basic usage documentation on the older versions is probably okay.

In any case, thanks for your job. Other than documentation, it worked for me :)

**Comment by 3v0k4** [comment]
@x-yuri just wanted to thank you. I'm using this instead of the main README all the time 🙂

**Comment by mhenrixon** [comment]
Closing — this is 5+ years old and the documentation has been significantly improved since. If you have specific suggestions for the current docs, please open a new issue.
